activerecord/lib/active_record/result.rb

```diff
@@ -98,11 +98,16 @@ def last
 
     def cast_values(type_overrides = {}) # :nodoc:
       types = columns.map { |name| column_type(name, type_overrides) }
-      result = rows.map do |values|
-        types.zip(values).map { |type, value| type.deserialize(value) }
+      length = columns.length
+      rows.each do |values|
```

Comment:
I think given the problem is pretty much `maps -> zip -> maps` here, right? I don't like that either.

However, I don't think we really need to track the index ourselves given we have `each_with_index`, so I'd like to propose this patch:

```ruby
    def cast_values!(type_overrides = {}) # :nodoc:
      rows.each do |values|
        columns.each_with_index do |column, index|
          values[index] = column_type(name, type_overrides).deserialize(values[index])
        end
      end
      columns.one? ? rows.map!(&:first) : rows
    end
```

I ran this on your branch to compare the number of objects and [seeing the same amount](https://gist.github.com/sikachu/24fd150f4e77654b52ad7a4f9e208df6), so I think this might be an alternative. What do you think?

Reply:
(Apparently I forgot to save the file, and so I was running the benchmark against wrong code ... anyway, I updated the result gist and the code suggestion above. Still imilar numbers of object count.)